ghostwriter/static/js/tinymce/plugins/ghostwriter-comments/plugin.min.js

```diff
@@ -0,0 +1,106 @@
+tinymce.PluginManager.add('ghostwriter-comments', function(editor) {
+    // Function to decode the users hex username into ascii
+    function hex2ascii(hexString) {
+        let decodedStr = '';
+        for (let i = 0; i < hexString.length; i += 2) {
+            const byte = parseInt(hexString.substr(i, 2), 16);
+            decodedStr += String.fromCharCode(byte);
+        }
+        return decodedStr
+    }
+
+    // Create Comment
+    function createComment(comment, selection) {
+        const commentNode = editor.selection.getNode();
+        let commentThread = readComments(commentNode);
+        commentThread.push({ author: username, comment: comment });
+        editor.execCommand('mceInsertContent', false, `<span class="comment" comment-data="${encodeURIComponent(JSON.stringify(commentThread))}">${selection}</span>`);
+    }
+
+    // Read Comments
+    function readComments(commentNode) {
+        let commentThread = [];
+        if (commentNode.nodeName === 'SPAN' && commentNode.hasAttribute('comment-data')) {
+            commentDataDecoded = decodeURIComponent(commentNode.getAttribute('comment-data'));
+            commentData = JSON.parse(commentDataDecoded);
+            commentData.forEach(item => {
+                const author = item.author;
+                const comment = item.comment;
+                commentThread.push({ author: author, comment: comment });
+            });
+        }
+        return commentThread;
+    }
+
+    // Update Comments
+
+    // Resolve Comments
+    function resolveComment(commentNode) {
+        commentNode.className = 'comment-resolved';
+    }
+
+    // Delete Comments
+
+
+    editor.ui.registry.addMenuItem('comment', {
+        icon: 'comment',
+        text: 'Comment',
+        onAction: function () {
+            const selection = editor.selection.getContent();
+            const node = editor.selection.getNode();
+            let commentsThread = '';
+            readComments(node).forEach(item => {
+                const author = item.author;
+                const comment = item.comment;
+                const float = author === username ? 'right' : 'left';
+                commentsThread +=
+                `<div id="comment-thread" style="float: ${float};">
```

Comment:
Since we're adding CSS to the CSS file above for avatars, we may as well provide the comment thread styles there too.

I recommend using [flexbox](https://css-tricks.com/snippets/css/a-guide-to-flexbox/) instead of clear here. Something like this, though you'll need to wrap the whole thread in a div:
```css
#entire-comment-thread {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
}
#entire-comment-thread .comment {
    margin-left: 0;
    margin-right: 20px;
}
#entire-comment-thread .comment.from-self {
    align-self: flex-end;
    margin-left: 20px;
    margin-right: 0;
}
```

(Also as-is, this code emits multiple elements with the same ID)

Reply:
Thanks, I agree using flexbox is better than clear. I have wrapped it all in the `entire-comment-thread` div and set the class of each comment depending on the author.
I am still unsure of the styling here...

Reply:
I've also added a border around each individual comment and positioned the avatar on the top right of this comment border